### ISSUE TITLE
Disable LBA weighting on files and SSDs - extend to SMR drives

### ISSUE THREAD

**Issue by Stoatwblr** [openzfs/zfs#10182]
This was originally  https://github.com/openzfs/zfs/commit/fb40095f5f0853946f8150481ca22602d1334dfe

Owing to the way that SMR drives work, LBA weighting is _extremely_ counterproductive. 

Unfortunately there are a lot of SMR drives in CMR clothing out there and the number is increasing - making things worse is that many don't report themselves as zoned devices.

However there's a ray of sunshine: Any rotational device which reports "trim" functionality is definitely SMR - or at least using zones internally and should not be LBA-weighted. (Example WDx0EFAX (RED) drives (EFRX are CMR)

Detecting SMR-but-not-trimmable drives is a bit harder (Example: Seagate ST3000DM003)

**Comment by behlendorf** [comment]
You can disable LBA weighting by setting the module option `metaslab_lba_weighting_enabled=0`.  If you find this to be beneficial for your devices we could consider adding some logic to `metaslab_space_weight()` which infers that rotational devices which support trim are almost certainly SMR devices.

**Comment by Stoatwblr** [comment]
Unless/until I can sucessufully get the drive to resilver and _stay_ in the array - see ticket #10214 and the recent block and files postings, I can't benchmark this.

However if you take the time to sit through Manfred Berger's presentation on SMR (OpenZFS forum, Paris 2015- https://www.youtube.com/watch?v=a2lnMxMUxyc) you'll understand why LBA weighting is utterly valueless on SMR drives - and quite likely to be harmful due to it causing drives to generate massively large indirect tables

Essentially a SMR zone is the mechanical equivalent of a SSD block and the drives translate every LBA you send them to something in the zones - meaning it doesn't matter where you _THINK_ you're putting data on the drive, the drive is making up its own mind about where it ends up. There is a complete disconnect between LBA space and the actual position on the platters.

Adding confusion the CMR (conventional(non-shingled) magnetic recording) landing space for writes can be analogised to TLC/QLC SSD's SLC write cache space - stuff sits there until the drive decides a final resting place during quiet periods (or until the CMR zone fills up)

**Comment by Stoatwblr** [comment]
After some more digging, I think LBA weighting is essentially valueless beyond the first few LBAs even on CMR drives.

The reason is because of the way sectors are addressed:

Explanation from 
https://github.com/westerndigitalcorporation/DiskSim/blob/master/STABLE/DiskSim_Linux_Generic_2.01.016/diskmodel/doc/dm_manual.pdf

1.4.2 G1 Layout
dm layout g1 LBN-to-PBN mapping scheme int required
-----
This specifies the type of LBN-to-PBN mapping used by the disk. 0 indicates that the conventional
mapping scheme is used: LBNs advance along the 0th track of the 0th cylinder, then along the
1st track of the 0th cylinder, thru the end of the 0th cylinder, then to the 0th track of the 1st
cylinder, and so forth. 1 indicates that the conventional mapping scheme is modified slightly, such
that cylinder switches do not involve head switches. Thus, after LBNs are assigned to the last track
of the 0th cylinder, they are assigned to the last track of the 1st cylinder, the next-to-last track of
the 1st cylinder, thru the 0th track of the 1st cylinder. LBNs are then assigned to the 0th track of
the 2nd cylinder, and so on (“first cylinder is normal”). 2 is like 1 except that the serpentine pattern
does not reset at the beginning of each zone; rather, even cylinders are always ascending and odd
cylinders are always descending.


In other words: LBA to platter/head allocation runs along the platters before switching heads and may (or may not) run in a serpentine manner to avoid the actuator having to seek from one extreme of the platter to the other when changing between heads on a sequential write/read

Which in turn means that LBA weighting is only valuable for the first N tracks of the first platter, as beyond that you have no idea what the speed will actually be. It may be that speed decreases, to a minimum with each track step and then increases to a peak before decreasing again (serpentine pattern), or it may decrease to a minimum, then snap back to peak speeds as the head switches to track0 on the next platter (traditional linear pattern)

**Comment by amotin** [comment]
While I have no objections that LBA weighting likely does not have much sense for SMR (the question is how to reliably detect that for device-managed SMR), as I see from the code, it is not really used for 3 years now, since 4e21fd060a567a9c8a1bd0e640985412181c1e33 switched weighting to SPACEMAP_HISTOGRAM.

**Comment by Stoatwblr** [comment]
In which case disabling or removing it entirely really should be a no-brainer

**Comment by amotin** [comment]
It may still be used for older pool, not upgraded to new features yet.  But since it should not affect new pools, I would not focus on it too much.  If there is a way to detect SMR drives (like rotating disk with TRIM support or something better), I see no problem to use it.  Otherwise I would leave it as-is

**Comment by ryao** [comment]
> Detecting SMR-but-not-trimmable drives is a bit harder (Example: Seagate ST3000DM003)

This is going to need a black list. However, even a blacklist is not reliable given Western Digital and others putting SMR into models that were previously PMR/CMR. :/

**Comment by jdrch** [comment]
> Seagate ST3000DM003

Are there more examples of such HDDs? Unless I'm mistaken, this seems to be a legacy product no longer made by Seagate. Unless there are more examples, this may just be an edge case worth documenting and writing a honking warning about, and not necessarily coding support for in the near term.

**Comment by Stoatwblr** [comment]
_ALL_ current Seagate Barracuda and Barracuda compute drives are undeclared DM-SMR without TRIM, Toshiba's drives don't declare TRIM either.

It really is a mess. Apart from the Hattis law class action in California over WD REDs, there's another case filed in New York over the other drives: https://classactionsreporter.com/wp-content/uploads/Western-Digital-SMR-Hard-Drives-Compl.pdf

**Comment by jdrch** [comment]
> _ALL_ current Seagate Barracuda and Barracuda compute drives are undeclared DM-SMR without TRIM

I don't think this is the case. I recently deployed a 2 TB Seagate Barracuda 2.5 in HDD that is SMR with TRIM. Here's the [output](https://www.reddit.com/r/DataHoarder/comments/hdad2l/deployed_a_seagate_barracuda_st2000lm015_2_tb_25/fvndz60/) of `# hdparm -I` for it.
